apollo-router/src/plugins/telemetry/apollo_otlp_exporter.rs

```diff
@@ -0,0 +1,323 @@
+    errors_configuration: ErrorsConfiguration,
+    include_span_names: HashSet<&'static str>,
+}
+
+impl ApolloOtlpExporter {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        endpoint: &Url,
+        protocol: &Protocol,
+        batch_config: &BatchProcessorConfig,
+        apollo_key: &str,
+        apollo_graph_ref: &str,
+        schema_id: &str,
+        errors_configuration: &ErrorsConfiguration,
+        include_span_names: HashSet<&'static str>,
+    ) -> Result<ApolloOtlpExporter, BoxError> {
+        tracing::debug!(endpoint = %endpoint, "creating Apollo OTLP traces exporter");
+
+        let mut metadata = MetadataMap::new();
+        metadata.insert("apollo.api.key", MetadataValue::try_from(apollo_key)?);
+        let otlp_exporter = match protocol {
+            Protocol::Grpc => {
+                let mut span_exporter = SpanExporterBuilder::from(
+                    opentelemetry_otlp::new_exporter()
+                        .tonic()
+                        .with_timeout(batch_config.max_export_timeout)
+                        .with_endpoint(endpoint.to_string())
+                        .with_metadata(metadata)
+                        .with_compression(opentelemetry_otlp::Compression::Gzip),
+                )
+                .build_span_exporter()?;
+
+                // This is a hack and won't be needed anymore once opentelemetry_otlp will be upgraded
+                span_exporter = if let opentelemetry_otlp::SpanExporter::Tonic {
+                    trace_exporter,
+                    metadata,
+                    timeout,
+                } = span_exporter
+                {
+                    opentelemetry_otlp::SpanExporter::Tonic {
+                        timeout,
+                        metadata,
+                        trace_exporter: trace_exporter.accept_compressed(CompressionEncoding::Gzip),
+                    }
+                } else {
+                    span_exporter
+                };
+
+                Arc::new(Mutex::new(span_exporter))
+            }
+            // So far only using HTTP path for testing - the Studio backend only accepts GRPC today.
+            Protocol::Http => Arc::new(Mutex::new(
+                SpanExporterBuilder::from(
+                    opentelemetry_otlp::new_exporter()
+                        .http()
+                        .with_timeout(batch_config.max_export_timeout)
+                        .with_endpoint(endpoint.to_string()),
+                )
+                .build_span_exporter()?,
+            )),
+        };
+
+        return Ok(Self {
+            endpoint: endpoint.clone(),
+            batch_config: batch_config.clone(),
+            apollo_key: apollo_key.to_string(),
+            resource_template: Resource::new([
+                KeyValue::new(
+                    "apollo.router.id",
+                    ROUTER_ID.get_or_init(Uuid::new_v4).to_string(),
```

Comment:
This is initialised in a couple of places now. Let's have a `fn router_id()` that contains `ROUTER_ID.get_or_init(||Uuid::new_v4().to_string())`

Reply:
SGTM 👍

Reply:
https://github.com/apollographql/router/pull/4982/commits/7d4a2aedb3ccf3f67ca923906fcc0af28af53b91